### PR TITLE
Parallelize exonize 12102023

### LINKED ISSUE #11: insert_fragments_calls
This relates to the functions `insert_fragments_calls()`  in sql_utils.py and `insert_fragments_table` in exonize_handler.py. 

- I was confused by the function `insert_fragments_calls()`. It returns two strings used to perform two SQL execute commands. I think it would make more sense to have functions for the two executes. 
- `insert_fragments_table`  has an internal function which in turn has an internal function. That is too many levels. The strand analysis could be done somewhere else, either when the blast results are collected in the first place or right before inserting into the DB tables.
- `get_gene_tuple()` returns a tuple with 6 elements. Element 3 is within parenthesis. Is there a reason for that? I printed the results and cannot see that it is needed.

### LINKED ISSUE #12: Refactor the blast code
I think clarity would be improved if Blast-related code was refactored into a separate file and class. The Exonize object is quite heavy with many attributes and methods. If blast code was moved to a separate class, it could be instantiated in exonize.py and all its gory details would be "hidden" there, before being shared as an attribute to Exonize. Ideally, the blast abstraction would be so good that Exonize does not even need to know how sequences are compared, but that is not a priority.

### LINKED ISSUE #20: Speed up - replace multithreading with multiprocessing

### LINKED ISSUE #24: Make a global var instead
https://github.com/msarrias/exonize/blob/cd5b0e1bafc51e066aef760b2745e1955913c8f7/exonize/exonize_handler.py#L63

### PR DESCRIPTION
**Pull request by msarrias**

### CODE REVIEW

**Comment by msarrias** [comment]
I tried running your code in the cluster, seems like it skipped the code you parallelized, was that expected? here you have part of the logs:

[slurm-41782695.txt](https://github.com/msarrias/exonize/files/12890071/slurm-41782695.txt)

The following error is also triggered with your code (and not with mine?)

> Runtime` error: error in view Filtered_full_length_events after drop column: no such column: f.percent_query

the percent_query column is in the Fragments table, which I used for creating the Filtered_full_length_events view.  My understanding is that once a view is created, it shouldn't be affected by any posterior modifications of the reference table.

Also, which version of Biopython and python3 are you using?

File: exonize/exonize.py [hunk 1]
```diff
@@ -80,9 +87,19 @@ def main():
...
                           threads=args.threads,
-                          timeout_db=args.timeout_db)
+                          timeout_db=args.timeout_db,
+                          genome_pickled_file_path=args.genome_pickled_file_path,
+                          )
     exonize_obj.run_exonize_pipeline()
 
 
 if __name__ == '__main__':
+    PROFILE_PATH = '/home/arthur/experimental/exonize/cProfile_dump_stats'
```
Comment (msarrias) [inline exonize/exonize.py]: :)

Comment (mvggz) [reply exonize/exonize.py]: That was a branch :), I didn't open a pull request with this inside. This is work in progress.

**Comment by mvggz** [comment]
> I tried running your code in the cluster, seems like it skipped the code you parallelized, was that expected? here you have part of the logs:

[slurm-41782695.txt](https://github.com/msarrias/exonize/files/12890071/slurm-41782695.txt)

The following error is also triggered with your code (and not with mine?)

    Runtime` error: error in view Filtered_full_length_events after drop column: no such column: f.percent_query

the percent_query column is in the Fragments table, which I used for creating the Filtered_full_length_events view. My understanding is that once a view is created, it shouldn't be affected by any posterior modifications of the reference table.

Also, which version of Biopython and python3 are you using?


I just ran `pip install -e .` from your repo, so it's the version specified in your `requirement.txt`. I ran it using 3.10.12 and 3.9.17, without much difference between the two of them.
From what I read of the logs, it's using the `os.fork`, but it's behaving very differently than on my local computer and with the tests data you sent me, can you send me the ones you used to produce that?

File: exonize/exonize_handler.py [hunk 2]
```diff
@@ -1240,34 +1337,105 @@ def run_exonize_pipeline(self) -> None:
...
+                # Benchmark with parallel computation using os.fork:
+                pr = cProfile.Profile()
+                pr.enable()
+                gc.collect()
+                gc.freeze()
+                transactions_pks: set[int]
+                status: int
+                code: int
+                forks: int = 0
+                FORKS_NUMBER = os.cpu_count()  # This is pretty greedy, could be changed and put in a config file
```
Comment (msarrias) [inline exonize/exonize_handler.py]: In terms of resource management, is this safe?

File: exonize/exonize_handler.py [hunk 3]
```diff
@@ -1,46 +1,85 @@
...
+    query_gene_ids_in_res_db,
+    query_fragments,
+    query_filtered_full_duplication_events,
+    query_concat_categ_pairs,
+    query_full_events,
+)
+from exonize.utils import (
+    get_interval_dictionary,
+    generate_unique_events_list,
+)
```
Comment (msarrias) [inline exonize/exonize_handler.py]: Why not import everything?

File: exonize/exonize_handler.py [hunk 4]
```diff
@@ -1240,34 +1337,105 @@ def run_exonize_pipeline(self) -> None:
...
+
+                # Benchmark with parallel computation using os.fork:
+                pr = cProfile.Profile()
+                pr.enable()
+                gc.collect()
+                gc.freeze()
+                transactions_pks: set[int]
+                status: int
+                code: int
+                forks: int = 0
```
Comment (msarrias) [inline exonize/exonize_handler.py]: Is this necessary, i.e. including the variable types? I have not done this throughout my code, should I?

File: exonize/exonize_handler.py [hunk 5]
```diff
@@ -1240,34 +1337,105 @@ def run_exonize_pipeline(self) -> None:
...
+                    # The parallel part happens because the main parent process will keep along the for loop, and will
+                    # fork more children, until the number of children reaches the maximum number of children allowed,
+                    # doing nothing else but forking until 'FORKS_NUMBER' is reached.
+                    if os.fork():
+                        forks += 1
+                        if forks >= FORKS_NUMBER:
+                            _, status = os.wait()
+                            code = os.waitstatus_to_exitcode(status)
+                            assert code in (os.EX_OK, os.EX_TEMPFAIL, os.EX_SOFTWARE)
+                            assert code != os.EX_SOFTWARE
```
Comment (msarrias) [inline exonize/exonize_handler.py:1410]: I don't see why including the  EX_SOFTWARE status in the first assert

**Pull request merged**